packages/world/ts/actions/callFrom.ts

```diff
@@ -0,0 +1,144 @@
+// Accepts either `worldFunctionToSystemFunction` or `publicClient`, but not both.
+type CallFromParameters = CallFromFunctionParameters | CallFromClientParameters;
+type CallFromBaseParameters = {
+  worldAddress: Hex;
+  delegatorAddress: Hex;
+};
+type CallFromFunctionParameters = CallFromBaseParameters & {
+  worldFunctionToSystemFunction: (worldFunctionSelector: Hex) => Promise<SystemFunction>;
+  publicClient?: never;
+};
+type CallFromClientParameters = CallFromBaseParameters & {
+  worldFunctionToSystemFunction?: never;
+  publicClient: PublicClient;
+};
+
+type SystemFunction = { systemId: Hex; systemFunctionSelector: Hex };
+
+// By extending viem clients with this function after delegation, the delegation is automatically applied to World contract writes.
+// This means that these writes are made on behalf of the delegator.
+// Internally, it transforms the write arguments to use `callFrom`.
+//
+// Accepts either `worldFunctionToSystemFunction` or `publicClient` as an argument.
+// `worldFunctionToSystemFunction` allows manually providing the mapping function, thus users can utilize their client store for the lookup.
+// If `publicClient` is provided instead, this function retrieves the corresponding system function from the World contract.
+//
+// The function mapping is cached to avoid redundant retrievals for the same World function.
+export function callFrom<TChain extends Chain, TAccount extends Account>(
+  params: CallFromParameters,
+): (client: WalletClient<Transport, TChain, TAccount>) => Pick<WalletActions<TChain, TAccount>, "writeContract"> {
+  return (client) => ({
+    // Applies to: `client.writeContract`, `getContract(client, ...).write`
+    writeContract: async (writeArgs): Promise<WriteContractReturnType> => {
+      // Skip if the contract isn't the World.
+      if (writeArgs.address !== params.worldAddress) {
+        return getAction(client, writeContract, "writeContract")(writeArgs);
+      }
+
+      // Encode the World's calldata (which includes the World's function selector).
+      const worldCalldata = encodeFunctionData({
+        abi: writeArgs.abi,
+        functionName: writeArgs.functionName,
+        args: writeArgs.args,
+      } as unknown as EncodeFunctionDataParameters);
+
+      // The first 4 bytes of calldata represent the function selector.
+      const worldFunctionSelector = slice(worldCalldata, 0, 4);
+
+      // Get the systemId and System's function selector.
+      const { systemId, systemFunctionSelector } = await worldFunctionToSystemFunction(params, worldFunctionSelector);
+
+      // Construct the System's calldata.
+      // If there's no args, use the System's function selector as calldata.
+      // Otherwise, use the World's calldata, replacing the World's function selector with the System's.
+      const systemCalldata =
+        worldCalldata === worldFunctionSelector
+          ? systemFunctionSelector
+          : concat([systemFunctionSelector, slice(worldCalldata, 4)]);
```

Comment:
why not always `concat([systemFunctionSelector, slice(worldCalldata, 4)]);`? if ` worldCalldata === worldFunctionSelector` then that means `slice(worldCalldata, 4)` is empty, so concat would be a noop, so we don't need to handle it with a special case?

Reply:
Using `const systemCalldata = concat([systemFunctionSelector, slice(worldCalldata, 4)]);` could result in an out-of-bounds error in `slice()` if there's no arg.
(`SliceOffsetOutOfBoundsError: Slice starting at offset "4" is out-of-bounds (size: 4).`)

I could alternatively use `size(worldCalldata) === 4` as a condition. Would that be better?

Reply:
could we use `readHex` from `@latticexyz/common` here? That should handle the case where the start would be out of bounds

Reply:
Maybe having a helper function that splits the function calldata into selector and argument data would simplify this function a bit.

Reply:
@alvrs I'll look into readHex. thanks!